### ISSUE TITLE
REST API: base URL returns 404, not clear how to find the endpoints

### ISSUE THREAD

**Issue by hongyi-zhao** [aiidateam/aiida-core#3916]
Hi,

I started the rest api as follows:

`$ verdi -p aiida restapi`

Then I open http://localhost:5000/ to request a full list of available endpoints for each resource, but failed with:
```

Not Found

The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.

```

Any hints?

Regards

**Comment by sphuber** [comment]
You need to add `api/v4` to the URL as marked in the top line of the output of `verdi restapi`:
```
 * REST API running on http://127.0.0.1:5000/api/v4
 * Serving Flask app "aiida.restapi.run_api" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```

The lower line is misleading indeed.

P.S.: these are typical questions that are better used for the mailing list. Please send  your questions there.

**Comment by hongyi-zhao** [comment]
@sphuber But I tried this: http://127.0.0.1:5000/api/v4
Still failed with:
```

$ curl -s http://127.0.0.1:5000/api/v4
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>404 Not Found</title>
<h1>Not Found</h1>
<p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>
```

P.S.:  Thanks for you hints, from now on, for the similar problems, I'll post on  the mailing list.

**Comment by sphuber** [comment]
what is the output when you run `verdi -p aiida restapi`?

**Comment by hongyi-zhao** [comment]
```
(aiida) werner@ubuntu-01:~$ verdi -p aiida restapi
 * REST API running on http://127.0.0.1:5000/api/v4
 * Serving Flask app "aiida.restapi.run_api" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

```

**Comment by sphuber** [comment]
To get the available endpoints do `curl -s http://127.0.0.1:5000/api/v4/server/endpoints/`

**Comment by hongyi-zhao** [comment]
Thanks, it does the trick, but not mentioned in the document.

Regards

**Comment by sphuber** [comment]
I agree, I will reopen this issue and rename it aptly so it can be fixed

**Comment by sphuber** [comment]
@asle85 @waychal @ltalirz there seems to be a bit of confusion on how to use the REST API. 

1. When launching `verdi restapi` there are two URLs printed, one with the correct suffix, one without. I think the latter (at the bottom) is automatically written by Flask, but it would be good if we can get rid of it.
2. Even when taking the correct URL `https:127.0.0.1:5000/api/v4` this will still raise a 404. It is not automatically clear what endpoints are available. Should we change the suggested URL to be `http://127.0.0.1:5000/api/v4/server/endpoints/`
3. We should probably update the [documentation](https://aiida-core.readthedocs.io/en/latest/restapi/index.html) to mention this endpoint, or have the base url automatically forward to the `server/endpoints` endpoint.

**Comment by ltalirz** [comment]
Thanks for the suggestions @sphuber ; I've mentioned exactly those before as well

> When launching verdi restapi there are two URLs printed, one with the correct suffix, one without. I think the latter (at the bottom) is automatically written by Flask, but it would be good if we can get rid of it.

I've looked into it now and unfortunately this is not so straightforward. 

We are using the built-in werkzeug server by flask and the hostname passed to the `run` function is also the one used to serve the API
https://github.com/pallets/werkzeug/blob/09a248649069780b4adfb910fb69f6fb304f2ea3/src/werkzeug/serving.py#L821-L831
Another way is to change the log level of the logger - I've tried that but I'm running into issues because of the internal reloading mechanism of the server, but for some reason passing `use_reloader=False` does not have the desired effect.
I've found *one* way to make it work, by setting `os.environ['WERKZEUG_RUN_MAIN'] = 'true'`
I'll try more when I find time.

> or have the base url automatically forward to the server/endpoints endpoint.

I think we should discuss this once in the AiiDA team. If we all agree on this, we should add it.
